File: DSA/LC/51_n-queens.cpp
#include <vector>
#include <string>
#include <unordered_set>

using namespace std;

class Solution
{
public:
    vector<vector<string>> solveNQueens(int n)
    {
        unordered_set<int> col;
        unordered_set<int> posDiag;
        unordered_set<int> negDiag;

        vector<vector<string>> res;
        vector<string> board(n, string(n, '.'));

        backtrack(0, n, col, posDiag, negDiag, board, res);
        return res;
    }

private:
    void backtrack(int r, int n, unordered_set<int> &col, unordered_set<int> &posDiag,
                   unordered_set<int> &negDiag, vector<string> &board,
                   vector<vector<string>> &res)
    {
        if (r == n)
        {
            res.push_back(board);
            return;
        }

        for (int c = 0; c < n; c++)
        {
            if (col.count(c) || posDiag.count(r + c) || negDiag.count(r - c))
            {
                continue;
            }

            col.insert(c);
            posDiag.insert(r + c);
            negDiag.insert(r - c);
            board[r][c] = 'Q';

            backtrack(r + 1, n, col, posDiag, negDiag, board, res);

            col.erase(c);
            posDiag.erase(r + c);
            negDiag.erase(r - c);
            board[r][c] = '.';
        }
    }
};

File: DSA/LC/51_n-queens.py
from typing import List


class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        solutions = []
        valid_chessboards = []
        queen_positions = []  # [(row, col)]
        forbidden_columns = set()

        def backtrack(row: int):
            if row == n:
                valid_chessboards.append(queen_positions.copy())
                return

            forbidden_diagonals = set()
            for row_index, col_index in queen_positions:
                row_diff = row - row_index
                forbidden_diagonals.add(col_index + row_diff)
                forbidden_diagonals.add(col_index - row_diff)

            for col in range(n):
                if col in forbidden_columns or col in forbidden_diagonals:
                    continue
                queen_positions.append((row, col))
                forbidden_columns.add(col)
                backtrack(row + 1)
                queen_positions.pop()
                forbidden_columns.remove(col)

        backtrack(0)

        for board in valid_chessboards:
            board_representation = [['.' for _ in range(n)] for _ in range(n)]
            for row, col in board:
                board_representation[row][col] = 'Q'
            solutions.append([''.join(row) for row in board_representation])

        return solutions
